kratos/utilities/data_type_traits.h

```diff
@@ -0,0 +1,642 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main author:     Suneth Warnakulasuriya
+//
+
+#pragma once
+
+// System includes
+#include <algorithm>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+// Project includes
+#include "containers/array_1d.h"
+#include "includes/ublas_interface.h"
+
+namespace Kratos {
+
+template<class TDataType> class DataTypeTraits
+{
+public:
+    ///@name Type definitions
+    ///@{
+
+    using ContainerType = TDataType;
+
+    using ValueType = TDataType;
+
+    using PrimitiveType = TDataType;
+
+    static constexpr bool HasContiguousPrimitiveData = true;
+
+    static constexpr bool HasDynamicMemoryAllocation = false;
+
+    ///@}
+    ///@name Public static operations
+    ///@{
+
+    static inline unsigned int Size(const ContainerType&)
+    {
+        return 1;
+    }
+
+    static inline std::vector<unsigned int> Shape(const ContainerType&)
+    {
+        return {};
+    }
+
+    static inline bool Reshape(
+        ContainerType&,
+        const std::vector<unsigned int>& rShape)
+    {
+        KRATOS_ERROR_IF(rShape.size() != 0)
+            << "Invalid shape given for primitive data type [ Expected shape = [], provided shape = "
+            << rShape << " ].\n";
+        return false;
+    }
+
+    inline static PrimitiveType const * GetContiguousData(const ContainerType& rValue)
+    {
+        return &rValue;
+    }
+
+    inline static PrimitiveType * GetContiguousData(ContainerType& rValue)
+    {
+        return &rValue;
+    }
+
+    inline static void FillToContiguousData(
+        PrimitiveType* pContiguousDataBegin,
+        const ContainerType& rValue)
+    {
+        *pContiguousDataBegin = rValue;
+    }
+
+    inline static void FillFromContiguousData(
+        ContainerType& rValue,
+        PrimitiveType const * pContiguousDataBegin)
+    {
+        rValue = *pContiguousDataBegin;
+    }
+
+    ///@}
+};
+
+template<class TDataType, std::size_t Dimension> class DataTypeTraits<array_1d<TDataType, Dimension>>
+{
+public:
+    ///@name Type definitions
+    ///@{
+
+    using ContainerType = array_1d<TDataType, Dimension>;
+
+    using ValueType = TDataType;
+
+    using PrimitiveType = typename DataTypeTraits<ValueType>::PrimitiveType;
+
+    static constexpr bool HasContiguousPrimitiveData = std::is_same_v<PrimitiveType, ValueType>;
+
+    static constexpr bool HasDynamicMemoryAllocation = DataTypeTraits<ValueType>::HasDynamicMemoryAllocation;
+
+    ///@}
+    ///@name Public static operations
+    ///@{
+
+    static inline unsigned int Size(const ContainerType& rContainer)
+    {
+        return Dimension * DataTypeTraits<ValueType>::Size(rContainer[0]);
```

Comment:
by the way, this approach won't work if your `ValueType` is dynamic, since each item can theoretically have a different size. This is true for all your `Size` implementations and any other function that checks the first value in the container and disregards the rest.

Reply:
done :)